mitmproxy/addons/export.py

```diff
@@ -18,36 +19,42 @@ def raise_if_missing_request(f: flow.Flow) -> None:
 
 def curl_command(f: flow.Flow) -> str:
     raise_if_missing_request(f)
-    data = "curl "
+    args = ["curl"]
     request = f.request.copy()  # type: ignore
     request.decode(strict=False)
     for k, v in request.headers.items(multi=True):
-        data += "-H '%s:%s' " % (k, v)
+        args += ["-H", "%s:%s" % (k, v)]
     if request.method != "GET":
-        data += "-X %s " % request.method
-    data += "'%s'" % request.url
+        args += ["-X", request.method]
+    args.append(request.url)
     if request.content:
-        data += " --data-binary '%s'" % strutils.bytes_to_escaped_str(
-            request.content,
-            escape_single_quotes=True
-        )
-    return data
+        try:
+            content = strutils.always_str(request.content)
+        except UnicodeDecodeError:
+            # shlex.quote doesn't support a bytes object
+            # see https://github.com/python/cpython/pull/10871
+            raise exceptions.CommandError("Request content must be valid unicode")
```

Comment:
I'm not really happy with this construct here - the idea of `strutils.bytes_to_escaped_str` is that it gives you something you can paste, even if it's binary. For example, if the request body is a null byte, you get a string with `\x00` back.

 How would you feel about `shlex.quote(strutils.bytes_to_escaped_str(request.content))`?

Reply:
Great find on the Bash feature! 
If I understand things correctly, we should `shlex.quote` all other parameters and then append binary content:
```python
command = ' '.join(shlex.quote(arg) for arg in args)
if request.content:
    command += (
        "--data-binary $'" + 
        strutils.bytes_to_escaped_str(request.content, escape_single_quotes=True) + 
        "'"
    )
return command
```

The problem with shlex.quote is that it would break the $ env:
```python
>>> print(shlex.quote(r"foo\55'\55"))
'foo\55'"'"'\55'
```

Makes sense?

(and we definitely need tests here)

Reply:
I'm not sure of how to proceed to support this. Calling `shlex.quote` on the output of `bytes_to_escaped_str` and prepending a `$` doesn't work properly based on the example you posted. Removing the `shlex.quote` call makes it work and it properly escapes single quotes, but that doesn't sound safe to me. We could also copy the code done in [this PR](https://github.com/python/cpython/pull/10871/files) that makes `shlex.quote` support bytes objects, but I don't know if that has been properly tested.

Another thing I noted is that none of this techniques will work properly if the body has a null byte, since the body is being passed as an argument, and [command arguments can't contain null bytes](https://eklitzke.org/unexpected-places-you-can-and-cant-use-null-bytes).

I think we should have a solution that is secure and that works in most cases. For the rest of the cases (like bodies that contain null bytes or that cannot be decoded to unicode) you can always use mitmproxy to replay requests. What do you think?

Reply:
> I think we should have a solution that is secure and that works in most cases. For the rest of the cases (like bodies that contain null bytes or that cannot be decoded to unicode) you can always use mitmproxy to replay requests. What do you think?

Fair enough - it's probably not a bad idea to not rely on  `bytes_to_escaped_str` for security.